runtime/codefrag.c

```diff
@@ -0,0 +1,126 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cambium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+/* A table of all code fragments (main program and dynlinked modules) */
+
+#include <string.h>
+#include <stddef.h>
+#include "caml/codefrag.h"
+#include "caml/misc.h"
+#include "caml/md5.h"
+#include "caml/memory.h"
+#include "caml/skiplist.h"
+
+static struct skiplist caml_code_fragments_list = SKIPLIST_STATIC_INITIALIZER;
+
+static struct ext_table caml_code_fragments_table = {0, 0, NULL};
+
+int caml_register_code_fragment(char * start, char * end,
+                                enum digest_status digest_kind,
+                                unsigned char * opt_digest)
+{
+  int pos;
+  struct code_fragment * cf = caml_stat_alloc(sizeof(struct code_fragment));
+
+  cf->code_start = start;
+  cf->code_end = end;
+  switch (digest_kind) {
+  case DIGEST_LATER:
+    break;
+  case DIGEST_NOW:
+    caml_md5_block(cf->digest, cf->code_start, cf->code_end - cf->code_start);
+    digest_kind = DIGEST_PROVIDED;
+    break;
+  case DIGEST_PROVIDED:
+    memcpy(cf->digest, opt_digest, 16);
+    break;
+  case DIGEST_IGNORE:
+    break;
+  }
+  cf->digest_status = digest_kind;
+  caml_skiplist_insert(&caml_code_fragments_list,
+                       (uintnat) start, (uintnat) cf);
+  if (caml_code_fragments_table.capacity == 0) {
+    caml_ext_table_init(&caml_code_fragments_table, 8);
+  }
+  pos = caml_ext_table_add(&caml_code_fragments_table, cf);
+  cf->fragnum = pos;
+  return pos;
+}
+
+void caml_remove_code_fragment(struct code_fragment * cf)
```

Comment:
When a code fragment is removed, the fragnums of all subsequent code fragments change. Won't this confuse the debugger, which still has the old fragnums in its `ev_frag`s?

Reply:
I was careful to reproduce exactly the behavior of the current code:
https://github.com/ocaml/ocaml/blob/2e4d1ecc12cfc6daa3b87e0400aed12544d877a9/runtime/meta.c#L156-L159

The effect of `caml_ext_table_remove` is to decrease the fragment numbers of all fragments "above" the one being removed.

I was hoping there was a corresponding adjustment to fragment numbers on the ocamldebug side, but I cannot find it anywhere in the ocamldebug sources.

@jhjourdan @whitequark : can you please clarify what should happen to fragment numbers when a fragment is unloaded?

Reply:
I unfortunately no longer recall OCaml-related specifics as I worked on this too long ago, but that being said, since the debugger is reliably notified of the lifetime of fragment numbers, IMO they can be safely reused.

Reply:
Thanks for your quick reply, however you may be missing the point.

Say we have 3 code fragments in use, with fragment numbers 0 1 2.

Say we unload the fragment number 1.

`caml_ext_table_remove` "slides" fragment number 2 at slot number 1.  So, from now on, the code that use to have fragment number 2 has fragment number 1.

The debugger is notified that fragment number 1 was unloaded.  However, I don't see where the debugger replaces fragment number 2 by fragment number 1.

Reply:
Oh, indeed I am missing the point. I suspect it's just broken; I remember that back when I worked on it, code loading was essentially nonfunctional in both ocamldebug and the emacs frontend.

Reply:
The issue isn't just ID reuse, it's that the IDs of code fragments change midway through their lifetimes as they get shifted around by other code fragments being removed.

Maybe we should use a skiplist rather than an ext_table to map fragment numbers to code_fragments? That way, we can give code fragments unique immutable IDs.

Reply:
If I understand correctly, code fragment numbers where important in the old world where we could only (easily) address PCs by knowing their code fragment. If we can now lookup PCs efficiently, do we still need the code fragment number in the debugger? (I suppose that there is no risk that PCs would overlap?)

Reply:
Recall that in the current implementation, code fragments can only be unloaded if they correspond to code written directly in the toplevel (and ony if this code does not include functions).

In this situation, it is very likely that the unloaded code fragment is the last one (well, except if this piece of code dynamically loads a plugin). 

So, the bug @xavierleroy is mentionning can only occur is very rare situations : when loading a plugin when running a piece of code from the toplevel, when this toplevel is watched by ocamldebug.

Now, if we want to avoid this, we can still assign code fragment unique identifiers by incrementing a counter each time a code fragment is loaded.

Reply:
All right.  I rewrote the code by 1- not changing fragment numbers, 2- reclaiming space in caml_code_fragments_table only when the removed fragment is the one most recently registered, and 3- leaving a hole in caml_code_fragments_table otherwise.

We could do better by maintaining a free list of holes in caml_code_fragments_table, to be reused at the next registration, or by switching to a skip list indexed by fragment numbers.  But this should do for now.

Reply:
> If we can now lookup PCs efficiently, do we still need the code fragment number in the debugger? (I suppose that there is no risk that PCs would overlap?)

We could use the start PC as code fragment identifier, I guess.  We would need to use nativeint instead of int in the debugger and in the protocol, though.

Reply:
More generally: I was trying to not change the debugger and the communication protocol, just improve performance of the "find fragment by PC" operation.